constchecker: check constants between Go/BPF

kkourt opened the pull request:
In the context of #12415, we wanted to add a test to compare values in
Go with ones in BPF. The first approach used `cgo`, which is not ideal.

This add some initial code to alignchecker that allows to check values
between BPF and go by reading the ELF .rodata section.

If this makes sense, it can be used for to add testing for #12415.

Signed-off-by: Kornilios Kourtis <kornilios@isovalent.com>

christarazi commented on pkg/alignchecker/alignchecker.go:
```diff
@@ -143,3 +145,86 @@ func check(name string, toCheck []reflect.Type, structs map[string]structInfo, c
+	}
+	defer f.Close()
+
+	elf_syms, err := f.Symbols()
```
Nit: In Golang, we generally [follow camel case for variable names](https://golang.org/doc/effective_go.html#mixed-caps). Plus, this will help keep the casing consistent in this function. :)

kkourt replied:
Thanks!

It seems that `golint` captures a lot of these, so hopefully I'll avoid them in the future.

christarazi commented on pkg/alignchecker/alignchecker.go:
```diff
@@ -143,3 +145,86 @@ func check(name string, toCheck []reflect.Type, structs map[string]structInfo, c
+
+		val_size := uint64(go_val.Type().Size())
+		if val_size != sym.Size {
+			return fmt.Errorf("%s: BPF size (%v) differs from Go value size (%v)", sym.Name, sym.Size, val_size)
```
This would a great candidate for a sentinel error value, where we create a specific error type representing this condition.

```
type ErrBPFSizeMismatch struct { # Feel free to pick a name you like better
     name string
     bpfSize, valueSize uint64
}

...

return ErrBPFSizeMismatch{
  name: sym.Name,
  bpfSize: sym.Size,
  valueSize: valSize
}
```

kkourt replied:
Thanks, makes sense.

christarazi commented on pkg/alignchecker/alignchecker.go:
```diff
@@ -143,3 +145,86 @@ func check(name string, toCheck []reflect.Type, structs map[string]structInfo, c
+
+	rodata := f.Section(".rodata")
+	if rodata == nil {
+		return fmt.Errorf("Failed to open .rodata section")
```
In Golang, we generally create errors that start with a lowercase. Also, if we are not formatting a message, then we want to just state what the error actually is (in form of a sentinel error variable). 

In this specific case, we indeed failed to open the `.rodata` section, but I presume it's because it's missing (?). So creating a sentinel error variable for this would be more ideal.

```
... (in global scope)
var ErrMissingROData = errors.New("missing .rodata")
...

rodata := f.Section(".rodata")
if rodata == nil {
	return ErrMissingROData
}
...
```

kkourt replied:
Thanks. I went a slightly different way, but I think it's in the same spirit.

christarazi commented on pkg/alignchecker/alignchecker_test.go:
```diff
@@ -119,3 +119,14 @@ func (t *AlignCheckerSuite) TestCheckStructAlignments(c *C) {
+		"bar_const": reflect.ValueOf(bar),
+	}
+	err := valueCheck(path, toCheck)
+	c.Assert(err, IsNil)
```
Can you add test cases for other conditions such as `toCheck` still having some elements left inside it and etc?

kkourt replied:
Thanks. Added test cases, and also updated the code to iterate over `toCheck` elements, instead of having to delete them and check if  it the map is empty at the end.

christarazi reviewed (changes requested):
Nice! Overall looks good, just have a few comments below.

aanm commented on pkg/alignchecker/alignchecker_test.go:
```diff
@@ -17,6 +17,7 @@
 package alignchecker
 
 import (
+	"github.com/cilium/cilium/pkg/checker"
```
The order of imports we are using is:

```
<golib>
<blank line>
<cilium imports>
<blank line>
<3rd party libs>
```

kkourt replied:
Thanks. Fixed.

tklauser commented on pkg/alignchecker/alignchecker.go:
```diff
@@ -143,3 +145,152 @@ func check(name string, toCheck []reflect.Type, structs map[string]structInfo, c
+}
+
+// valueCheck reads ELF symbols from .rodata and compare the values against given toCheck values.
+// It returns  the first error it encouters, or nil if succesful.
```
Nit:

```suggestion
// It returns  the first error it encounters, or nil if succesful.
```

tklauser commented on pkg/alignchecker/alignchecker.go:
```diff
@@ -143,3 +145,152 @@ func check(name string, toCheck []reflect.Type, structs map[string]structInfo, c
+		return &ErrElf{"no symbols"}
+	}
+
+	elfSymsMap := map[string]elf.Symbol{}
```
Using the `make` builtin to create empty maps is more idiomatic and would also allow to specify the expected size:

```suggestion
	elfSymsMap := make(map[string]elf.Symbol, len(elfSyms))
```

tklauser commented on pkg/alignchecker/alignchecker.go:
```diff
@@ -143,3 +145,152 @@ func check(name string, toCheck []reflect.Type, structs map[string]structInfo, c
+		// case reflect.Int8:
+		// case reflect.Int16:
+		// case reflect.Int32:
+		// case reflect.Int64:
```
Why are these commented out? Don't we need to handle them? Might be worth to explain why or just drop the commented out lines if they're not needed.

Same for `reflect.Uint{16,32}` below.

kkourt replied:
We currently only need the `Uint8` case. I added the `Unit64`, just to make sure that the code would work.

The commented-out types are just scaffolding in case we want to add different types.
I think it's useful to keep them, so I will add a comment to explain their purpose.

tklauser reviewed (changes requested):
Nice work.

Some small comments inline, nothing major. Probably not worth to respin the PR if these are the only requested changes.

kkourt committed:
loadbalancer: check consts between Go and BPF

The goal is to ensure that common constants between Go and BPF match.

This commit adds:
 - constchecker, which checks values between BPF and Go by reading the
 ELF .rodata section of a BPF object file.
 - two tests, one for constchecker itself with dummy constants and one
 for loadbalancer.

Signed-off-by: Kornilios Kourtis <kornilios@isovalent.com>

tklauser reviewed (approved):
Still LGTM with the `pkg/constchecker` rename. Thanks.

aanm commented on pkg/constchecker/testdata/Makefile:
```diff
@@ -0,0 +1,11 @@
+
+bpf_foo.o: bpf_foo.ll
+	llc -march=bpf -mcpu=probe -mattr=dwarfris -filetype=obj -o bpf_foo.o bpf_foo.ll
+
```
When is these targets executed? Can't we make this part of the release tests? Do we need to include the `bpf_foo.o` into the tree?

kkourt replied:
I think it's simpler if we just keep the `bpf_foo.o` in the tree. It is not supposed to change, unless we add more tests (e.g., if we want to support additional types). The `Makefile` was just for my convenience during development.

borkmann replied:
I think if it's not too much complexity I would prefer if we don't check-in a precompiled .o file, also as we might upgrade our internal clang/llc build for the Cilium image (https://github.com/cilium/image-tools/tree/master/images/llvm), so it would be good to then build bpf_foo.c with the same compiler binaries which also build Cilium's BPF programs.

kkourt replied:
OK, I removed the .o file. Instead, I pass the object file to the constchecker test via an environment variable, which is what I also do for the loadbalancer test.

aanm commented on pkg/constchecker/testdata/bpf_foo.c:
```diff
@@ -0,0 +1,30 @@
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
```
```
// SPDX-License-Identifier: GPL-2.0
/* Copyright (C) 2020 Authors of Cilium */
```

kkourt replied:
Thanks.

aanm commented on test/bpf/check-consts.sh:
```diff
@@ -0,0 +1,36 @@
+# NB: in both cases, we pass proper flags to go test, so that it only executes
+# the constchecker tests
+ CILIUM_CONSTCHECKER_BPF_OBJ=$BPFDIR/bpf_constchecker_self.o \
+	 go test github.com/cilium/cilium/pkg/constchecker -run '^Test$' -check.f 'ConstSuite.*'
```
I _think_ there are Makefle environment variables set for `go` and `test`. I know this is a script but I'm wondering if it would make sense to keep them consistent.

kkourt replied:
Indeed, they seem to be. My original thought was to add this test in the `Makefile` (defining the `CILIUM_CONSTCHECKER_BPF_OBJ` env variable would be enough), but this felt too intrusive/complex, so I decided to put it on a separate script. Since the const are fairly simple and have no other dependencies except for `cosntchecker`, I would think that running them as they are would be OK.

nebril reviewed (approved):
Test bits look good to me.

joestringer commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -56,14 +56,14 @@ const (
+	serviceFlagHostPort        = ServiceFlags(8)
+	serviceFlagSessionAffinity = ServiceFlags(16)
+	serviceFlagLoadBalancer    = ServiceFlags(32)
+	serviceFlagRoutable        = ServiceFlags(64)
```
nit: These seem like they would benefit from using hexadecimal representation since the bit values are more important than the numbers.

joestringer commented on pkg/loadbalancer/loadbalancer_test.go:
```diff
@@ -432,3 +435,26 @@ func TestServiceFlags_String(t *testing.T) {
+//
+// C struct size info is extracted from the given ELF object file .rodata section.
+func (t *ConstSuite) TestConst(c *check.C) {
+	// Validate alignments of C and Go equivalent structs
```
nit: This comment talks about alignments and structs, but the constchecker doesn't appear to deal with these things.

joestringer commented on test/bpf/check-consts.sh:
```diff
@@ -0,0 +1,36 @@
+# the constchecker tests
+ CILIUM_CONSTCHECKER_BPF_OBJ=$BPFDIR/bpf_constchecker_self.o \
+	 go test github.com/cilium/cilium/pkg/constchecker -run '^Test$' -check.f 'ConstSuite.*'
+ CILIUM_CONSTCHECKER_BPF_OBJ=$BPFDIR/bpf_constchecker.o \
+	 go test github.com/cilium/cilium/pkg/loadbalancer -run '^Test$' -check.f 'ConstSuite.*'
```
Will these tests be run twice now, once via the standard `make unit-tests` and once via this script?

This shouldn't be a big deal as I doubt these tests take very long at all but it's more of a question around divergence in how the unit tests are run.

kkourt replied:
They will not run twice. If `CILIUM_CONSTCHECK_BPF` is not defined (as is the case for `make unit-tests`), then the constant checks will be skipped.

joestringer reviewed (changes requested):
This seems like a reasonable way to perform such checks. At a high level, I wonder whether we would benefit from moving more towards a model where the Golang implementation generates all of these constants from the Golang definition into a header file instead, then the BPF code simply includes those headers. That way, the constants would be ensured to be the same without having to perform additional checks afterwards.

That said, modulo the minor nits below I'm fine with the current approach.

joestringer reviewed (approved):
I'm fine with the changes. We still have a partially-finished discussion above about the overall approach which could end in either deciding to merge this PR or not, so my approval is subject to that. I appreciate the effort looking into this either way :-)
